server/users.go

```diff
@@ -1,317 +1,153 @@
+	Provider string    `json:"provider,omitempty"`
+	Scheme   string    `json:"scheme,omitempty"`
+}
 
-	res := newUserResponse(srcID, u.Name).WithPermissions(u.Permissions)
-	if _, hasRoles := h.hasRoles(ctx, ts); hasRoles {
-		res.WithRoles(srcID, u.Roles)
+func newUserResponse(u *chronograf.User) *userResponse {
+	return &userResponse{
+		ID:       u.ID,
+		Name:     u.Name,
+		Provider: u.Provider,
+		Scheme:   u.Scheme,
+		Links: selfLinks{
+			Self: fmt.Sprintf("/chronograf/v1/users/%v", u.ID),
+		},
 	}
-	encodeJSON(w, http.StatusOK, res, h.Logger)
 }
 
-// RemoveSourceUser removes the user from the InfluxDB source
-func (h *Service) RemoveSourceUser(w http.ResponseWriter, r *http.Request) {
+func (s *Service) UserID(w http.ResponseWriter, r *http.Request) {
 	ctx := r.Context()
-	uid := httprouter.GetParamFromContext(ctx, "uid")
 
-	_, store, err := h.sourceUsersStore(ctx, w, r)
+	id := httprouter.GetParamFromContext(ctx, "id")
+	user, err := s.UsersStore.Get(ctx, id)
 	if err != nil {
+		Error(w, http.StatusBadRequest, err.Error(), s.Logger)
 		return
 	}
 
-	if err := store.Delete(ctx, &chronograf.User{Name: uid}); err != nil {
-		Error(w, http.StatusBadRequest, err.Error(), h.Logger)
-		return
-	}
-
-	w.WriteHeader(http.StatusNoContent)
+	res := newUserResponse(user)
+	encodeJSON(w, http.StatusOK, res, s.Logger)
 }
 
-// UpdateSourceUser changes the password or permissions of a source user
-func (h *Service) UpdateSourceUser(w http.ResponseWriter, r *http.Request) {
+func (s *Service) NewUser(w http.ResponseWriter, r *http.Request) {
 	var req userRequest
 	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
-		invalidJSON(w, h.Logger)
-		return
-	}
-	if err := req.ValidUpdate(); err != nil {
-		invalidData(w, err, h.Logger)
+		invalidJSON(w, s.Logger)
 		return
 	}
 
-	ctx := r.Context()
-	uid := httprouter.GetParamFromContext(ctx, "uid")
-	srcID, ts, err := h.sourcesSeries(ctx, w, r)
-	if err != nil {
+	if err := req.ValidCreate(); err != nil {
+		invalidData(w, err, s.Logger)
 		return
 	}
 
+	ctx := r.Context()
 	user := &chronograf.User{
-		Name:        uid,
-		Passwd:      req.Password,
-		Permissions: req.Permissions,
-		Roles:       req.Roles,
-	}
-	store := ts.Users(ctx)
-
-	if err := store.Update(ctx, user); err != nil {
-		Error(w, http.StatusBadRequest, err.Error(), h.Logger)
-		return
+		Name:     req.Name,
+		Provider: req.Provider,
+		Scheme:   req.Scheme,
 	}
 
-	u, err := store.Get(ctx, uid)
+	res, err := s.UsersStore.Add(ctx, user)
 	if err != nil {
-		Error(w, http.StatusBadRequest, err.Error(), h.Logger)
+		Error(w, http.StatusBadRequest, err.Error(), s.Logger)
 		return
 	}
 
-	res := newUserResponse(srcID, u.Name).WithPermissions(u.Permissions)
-	if _, hasRoles := h.hasRoles(ctx, ts); hasRoles {
-		res.WithRoles(srcID, u.Roles)
-	}
-	w.Header().Add("Location", res.Links.Self)
-	encodeJSON(w, http.StatusOK, res, h.Logger)
+	cu := newUserResponse(res)
+	w.Header().Add("Location", cu.Links.Self)
```

Comment:
While I haven't done this before (I should have) would you make adding a Location to the header a helper function?

Reply:
such as `addLocationHeader(w.Header(), cu.links.self)` ? would you mind explaining your thoughts on the motivation for this?

Reply:
```go
func location(w http.ResponseWriter, self string){
    w.Header().Add("Location", self)
}
```

In the chronograf API I've not been good at remembering to add the Location header where it is needed.  A helper function would really help to make sure we do it in all places (POSTs and PATCHes)